File: storeroom/dataset/pipeline.py
import numpy as np
import pandas as pd
from scipy import signal
from scipy.signal import find_peaks, periodogram, welch
from scipy.stats import kurtosis, skew, chi2


def load_and_preprocess_data(dataframe, fill_method='ffill'):
    """
    Load and preprocess the raw sensor data.
    """
    dataframe.fillna(method=fill_method, inplace=True)
    return dataframe


def normalize_and_remove_outliers(data):
    """
    Normalize sensor data using Z-score and remove the top and bottom 1% of values.
    """
    normalized_data = (data - data.mean()) / data.std()
    lower_bound = normalized_data.quantile(0.01)
    upper_bound = normalized_data.quantile(0.99)
    filtered = normalized_data.where((normalized_data > lower_bound) & (normalized_data < upper_bound))
    return filtered.dropna()


def segment_data(data, window_size_seconds=5):
    """
    Segment data into fixed-size windows based on the timestamp.
    """
    data['timestamp'] = pd.to_datetime(data['timestamp'], errors='coerce')
    start_time = data['timestamp'].min()
    data['seconds_since_start'] = (data['timestamp'] - start_time).dt.total_seconds()
    data['window_id'] = (data['seconds_since_start'] // window_size_seconds).astype(int)
    return data


def calculate_confidence_ellipse_area(x, y, confidence=0.95):
    """
    Calculate the area of the ellipse that encloses the 95% confidence interval
    of the 2D points (x, y). For a bivariate normal distribution, the area is:
      A = π * (chi2.ppf(confidence, 2)) * sqrt(det(Cov))
    """
    data = np.vstack([x, y])
    cov = np.cov(data)
    det_cov = np.linalg.det(cov)
    chi2_val = chi2.ppf(confidence, df=2)
    area = np.pi * chi2_val * np.sqrt(det_cov)
    return area


def calculate_sway_volume(data_x, data_y, data_z, confidence=0.95):
    """
    Calculate the volume of the sphere that encloses 95% of the 3D points.
    """
    points = np.vstack([data_x, data_y, data_z]).T
    center = np.mean(points, axis=0)
    distances = np.linalg.norm(points - center, axis=1)
    r = np.percentile(distances, 95)
    volume = (4/3) * np.pi * (r ** 3)
    return volume


def calculate_gait_features(acc_data, frequency=50):
    """
    Calculate gait features: number of steps and cadence.
    Returns (steps, cadence, peaks).
    """
    # Compute magnitude of accelerometer data (assumed gravity-corrected)
    g_mag = np.sqrt(np.sum(np.square(acc_data), axis=1))
    threshold = np.mean(g_mag) + np.std(g_mag)
    peaks, _ = find_peaks(g_mag, height=threshold)
    steps = len(peaks)
    duration_minutes = (len(acc_data) / frequency) / 60
    cadence = steps / duration_minutes if duration_minutes > 0 else 0
    return steps, cadence, peaks


def calculate_statistical_features(data):
    """
    Compute skewness and kurtosis of the data.
    """
    return skew(data), kurtosis(data)


def calculate_harmonics(data):
    """
    Calculate Total Harmonic Distortion (THD) using a periodogram.
    """
    freqs, pxx = periodogram(data)
    if len(pxx) < 6 or pxx[1] == 0:
        return 0
    fundamental = np.abs(pxx[1])
    harmonics = np.abs(pxx[2:6])
    thd = np.sqrt(np.sum(harmonics**2)) / fundamental
    return thd


def calculate_velocity_and_residual(acc_signal, frequency=50):
    """
    Detrend and integrate a gravity-corrected acceleration signal to obtain a velocity signal,
    then compute the average gait velocity and residual step length from detected steps.
    """
    dt = 1 / frequency
    acc_detrended = signal.detrend(acc_signal)
    velocity = np.cumsum(acc_detrended) * dt

    peaks, _ = find_peaks(acc_signal, height=np.mean(acc_signal) + np.std(acc_signal))
    if len(peaks) < 2:
        return 0, 0  # Not enough steps to compute features

    step_lengths = []
    for i in range(1, len(peaks)):
        start, end = peaks[i-1], peaks[i]
        displacement = np.trapz(velocity[start:end], dx=dt)
        step_lengths.append(displacement)
    avg_step_length = np.mean(step_lengths)
    residual_step_length = np.mean(np.abs(np.array(step_lengths) - avg_step_length))
    total_displacement = np.trapz(velocity, dx=dt)
    total_time = len(acc_signal) * dt
    avg_velocity = total_displacement / total_time if total_time > 0 else 0
    return avg_velocity, residual_step_length


def calculate_step_time_features(acc_signal, frequency=50):
    """
    Calculate average step time and residual step time from the accelerometer signal.
    """
    dt = 1 / frequency
    peaks, _ = find_peaks(acc_signal, height=np.mean(acc_signal) + np.std(acc_signal))
    if len(peaks) < 2:
        return 0, 0
    step_intervals = np.diff(peaks) * dt
    avg_step_time = np.mean(step_intervals)
    residual_step_time = np.mean(np.abs(step_intervals - avg_step_time))
    return avg_step_time, residual_step_time


def calculate_velocity_features(acc_data, frequency=50):
    """
    Calculate mean and variance for velocity signals obtained from accelerometer data.
    Expects a DataFrame with columns: 'acc_x', 'acc_y', 'acc_z'.
    """
    dt = 1 / frequency
    velocity_x = np.cumsum(acc_data['acc_x'] * dt)
    velocity_y = np.cumsum(acc_data['acc_y'] * dt)
    velocity_z = np.cumsum(acc_data['acc_z'] * dt)
    
    return (np.mean(velocity_x), np.var(velocity_x),
            np.mean(velocity_y), np.var(velocity_y),
            np.mean(velocity_z), np.var(velocity_z))


def calculate_angular_velocity_features(gyro_data, frequency=50):
    """
    Calculate mean and variance for angular velocity signals.
    Expects a DataFrame with columns: 'gyro_x', 'gyro_y', 'gyro_z'.
    """
    dt = 1 / frequency
    angular_velocity_x = np.cumsum(gyro_data['gyro_x'] * dt)
    angular_velocity_y = np.cumsum(gyro_data['gyro_y'] * dt)
    angular_velocity_z = np.cumsum(gyro_data['gyro_z'] * dt)
    
    return (np.mean(angular_velocity_x), np.var(angular_velocity_x),
            np.mean(angular_velocity_y), np.var(angular_velocity_y),
            np.mean(angular_velocity_z), np.var(angular_velocity_z))


def calculate_combined_features(df):
    """
    Calculate all high-level features for both smartphone and smartwatch data.
    """
    # Smartphone features
    smartphone_acc = df[['acc_x_smartphone', 'acc_y_smartphone', 'acc_z_smartphone']].values
    steps, cadence, _ = calculate_gait_features(smartphone_acc, frequency=50)
    avg_velocity, residual_step_length = calculate_velocity_and_residual(df['acc_z_smartphone'].values, frequency=50)
    avg_step_time, residual_step_time = calculate_step_time_features(df['acc_z_smartphone'].values, frequency=50)
    
    smartphone_acc_df = df[['acc_x_smartphone', 'acc_y_smartphone', 'acc_z_smartphone']].copy()
    smartphone_acc_df.columns = ['acc_x', 'acc_y', 'acc_z']
    velocity_feats = calculate_velocity_features(smartphone_acc_df, frequency=50)
    
    XY_sway_area = calculate_confidence_ellipse_area(df['gyro_x_smartphone'], df['gyro_y_smartphone'])
    YZ_sway_area = calculate_confidence_ellipse_area(df['gyro_y_smartphone'], df['gyro_z_smartphone'])
    XZ_sway_area = calculate_confidence_ellipse_area(df['gyro_x_smartphone'], df['gyro_z_smartphone'])
    sway_volume = calculate_sway_volume(df['gyro_x_smartphone'], df['gyro_y_smartphone'], df['gyro_z_smartphone'])
    
    freq_ratio = calculate_frequency_ratio(df['acc_z_smartphone'])
    band_power = calculate_band_power(df['acc_z_smartphone'])
    snr = calculate_signal_noise_ratio(df['acc_z_smartphone'])
    skewness, kurt = calculate_statistical_features(df['acc_z_smartphone'])
    thd = calculate_harmonics(df['acc_z_smartphone'])
    
    # Smartwatch features (accelerometer velocity and gyroscope angular velocity)
    smartwatch_acc = df[['acc_x_smartwatch', 'acc_y_smartwatch', 'acc_z_smartwatch']].copy()
    smartwatch_acc.columns = ['acc_x', 'acc_y', 'acc_z']
    velocity_feats_sw = calculate_velocity_features(smartwatch_acc, frequency=50)
    
    smartwatch_gyro = df[['gyro_x_smartwatch', 'gyro_y_smartwatch', 'gyro_z_smartwatch']].copy()
    smartwatch_gyro.columns = ['gyro_x', 'gyro_y', 'gyro_z']
    angular_velocity_feats = calculate_angular_velocity_features(smartwatch_gyro, frequency=50)
    
    XY_sway_area_sw = calculate_confidence_ellipse_area(df['gyro_x_smartwatch'], df['gyro_y_smartwatch'])
    YZ_sway_area_sw = calculate_confidence_ellipse_area(df['gyro_y_smartwatch'], df['gyro_z_smartwatch'])
    XZ_sway_area_sw = calculate_confidence_ellipse_area(df['gyro_x_smartwatch'], df['gyro_z_smartwatch'])
    sway_volume_sw = calculate_sway_volume(df['gyro_x_smartwatch'], df['gyro_y_smartwatch'], df['gyro_z_smartwatch'])
    
    return pd.Series({
        # Smartphone features
        'steps_smartphone': steps,
        'cadence_smartphone': cadence,
        'avg_velocity_smartphone': avg_velocity,
        'residual_step_length_smartphone': residual_step_length,
        'avg_step_time_smartphone': avg_step_time,
        'residual_step_time_smartphone': residual_step_time,
        'XY_sway_area_smartphone': XY_sway_area,
        'YZ_sway_area_smartphone': YZ_sway_area,
        'XZ_sway_area_smartphone': XZ_sway_area,
        'sway_volume_smartphone': sway_volume,
        'frequency_ratio_smartphone': freq_ratio,
        'band_power_smartphone': band_power,
        'signal_noise_ratio_smartphone': snr,
        'skewness_smartphone': skewness,
        'kurtosis_smartphone': kurt,
        'total_harmonic_distortion_smartphone': thd,
        'velocity_mean_X_smartphone': velocity_feats[0],
        'velocity_variance_X_smartphone': velocity_feats[1],
        'velocity_mean_Y_smartphone': velocity_feats[2],
        'velocity_variance_Y_smartphone': velocity_feats[3],
        'velocity_mean_Z_smartphone': velocity_feats[4],
        'velocity_variance_Z_smartphone': velocity_feats[5],
        # Smartwatch features
        'XY_sway_area_smartwatch': XY_sway_area_sw,
        'YZ_sway_area_smartwatch': YZ_sway_area_sw,
        'XZ_sway_area_smartwatch': XZ_sway_area_sw,
        'sway_volume_smartwatch': sway_volume_sw,
        'velocity_mean_X_smartwatch': velocity_feats_sw[0],
        'velocity_variance_X_smartwatch': velocity_feats_sw[1],
        'velocity_mean_Y_smartwatch': velocity_feats_sw[2],
        'velocity_variance_Y_smartwatch': velocity_feats_sw[3],
        'velocity_mean_Z_smartwatch': velocity_feats_sw[4],
        'velocity_variance_Z_smartwatch': velocity_feats_sw[5],
        'angular_velocity_mean_X_smartwatch': angular_velocity_feats[0],
        'angular_velocity_variance_X_smartwatch': angular_velocity_feats[1],
        'angular_velocity_mean_Y_smartwatch': angular_velocity_feats[2],
        'angular_velocity_variance_Y_smartwatch': angular_velocity_feats[3],
        'angular_velocity_mean_Z_smartwatch': angular_velocity_feats[4],
        'angular_velocity_variance_Z_smartwatch': angular_velocity_feats[5]
    })


def calculate_features(data):
    """
    Apply feature extraction on each window of segmented data.
    """
    return data.groupby('window_id').apply(calculate_combined_features).reset_index()


def sensor_data_pipeline(raw_data):
    """
    Run the complete ETL pipeline: preprocessing, outlier removal, segmentation,
    and feature extraction.
    """
    preprocessed = load_and_preprocess_data(raw_data)
    cleaned = normalize_and_remove_outliers(preprocessed)
    segmented = segment_data(cleaned)
    features = calculate_features(segmented)
    return features

File: storeroom/dataset/pipeline.dart
import 'dart:math';
import 'dart:convert';

/// Utility functions

double mean(List<double> values) =>
    values.isEmpty ? 0 : values.reduce((a, b) => a + b) / values.length;

double std(List<double> values) {
  double m = mean(values);
  double sumSq = values.map((v) => pow(v - m, 2)).reduce((a, b) => a + b);
  return sqrt(sumSq / values.length);
}

/// Computes the qth quantile (q between 0 and 1) of a sorted list.
double quantile(List<double> sortedValues, double q) {
  if (sortedValues.isEmpty) return 0;
  int index = ((sortedValues.length - 1) * q).round();
  return sortedValues[index];
}

/// Returns the covariance of two lists.
double covariance(List<double> xs, List<double> ys) {
  if (xs.length != ys.length || xs.isEmpty) return 0;
  double mX = mean(xs);
  double mY = mean(ys);
  double sum = 0;
  for (int i = 0; i < xs.length; i++) {
    sum += (xs[i] - mX) * (ys[i] - mY);
  }
  return sum / xs.length;
}

/// Returns the determinant of a 2x2 matrix.
double det2x2(double a, double b, double c, double d) => a * d - b * c;

//
// Data structures and basic pipeline functions
//

/// In our case, we assume sensor data is provided as a List of Maps,
/// where each Map represents a row with keys like 'timestamp', 'acc_x_smartphone', etc.
typedef SensorRow = Map<String, dynamic>;

/// Fills missing values using forward fill.
List<SensorRow> loadAndPreprocessData(List<SensorRow> data, {String fillMethod = 'ffill'}) {
  // Simple forward fill implementation (assumes keys are present in all rows).
  Map<String, dynamic> lastSeen = {};
  for (var row in data) {
    row.forEach((key, value) {
      if (value == null && lastSeen.containsKey(key)) {
        row[key] = lastSeen[key];
      } else if (value != null) {
        lastSeen[key] = value;
      }
    });
  }
  return data;
}

/// Normalizes numeric columns (Z-score) and removes rows with any value below the 1st
/// or above the 99th percentile.
List<SensorRow> normalizeAndRemoveOutliers(List<SensorRow> data) {
  if (data.isEmpty) return data;
  // Get list of numeric keys (exclude timestamp)
  List<String> keys = data.first.keys.where((k) => k != 'timestamp').toList();
  // Compute mean and std for each key
  Map<String, double> means = {};
  Map<String, double> stds = {};
  for (var key in keys) {
    List<double> col =
        data.map((row) => (row[key] as num).toDouble()).toList();
    means[key] = mean(col);
    stds[key] = std(col);
  }
  // Normalize each row
  List<SensorRow> normalized = [];
  for (var row in data) {
    SensorRow newRow = Map.from(row);
    bool valid = true;
    for (var key in keys) {
      double value = (row[key] as num).toDouble();
      double z = stds[key]! > 0 ? (value - means[key]!) / stds[key]! : 0;
      newRow[key] = z;
    }
    normalized.add(newRow);
  }
  // For each column, compute 1% and 99% quantile over normalized data.
  Map<String, double> lowerBounds = {};
  Map<String, double> upperBounds = {};
  for (var key in keys) {
    List<double> col =
        normalized.map((row) => row[key] as double).toList()..sort();
    lowerBounds[key] = quantile(col, 0.01);
    upperBounds[key] = quantile(col, 0.99);
  }
  // Filter out rows that have any column outside bounds.
  return normalized.where((row) {
    for (var key in keys) {
      double value = row[key] as double;
      if (value < lowerBounds[key]! || value > upperBounds[key]!) return false;
    }
    return true;
  }).toList();
}

/// Segments data into windows of fixed seconds (adds 'window_id' field).
List<SensorRow> segmentData(List<SensorRow> data, {int windowSizeSeconds = 5}) {
  if (data.isEmpty) return data;
  // Convert timestamps and compute seconds since start.
  DateTime startTime = DateTime.parse(data.first['timestamp']);
  for (var row in data) {
    DateTime ts = DateTime.parse(row['timestamp']);
    double secondsSinceStart = ts.difference(startTime).inMilliseconds / 1000.0;
    row['seconds_since_start'] = secondsSinceStart;
    row['window_id'] = (secondsSinceStart ~/ windowSizeSeconds);
  }
  return data;
}

//
// Feature computation functions
//

double calculateConfidenceEllipseArea(List<double> x, List<double> y, {double confidence = 0.95}) {
  double covXY = covariance(x, y);
  double covXX = covariance(x, x);
  double covYY = covariance(y, y);
  double det = det2x2(covXX, covXY, covXY, covYY);
  // For chi-square quantile, approximate with fixed value for 2 df and 95% confidence.
  // chi2.ppf(0.95, 2) ≈ 5.991
  double chi2Val = 5.991;
  return pi * chi2Val * sqrt(det.abs());
}

double calculateSwayVolume(List<double> x, List<double> y, List<double> z, {double confidence = 0.95}) {
  int n = x.length;
  List<List<double>> points = List.generate(n, (i) => [x[i], y[i], z[i]]);
  // Compute mean (center)
  List<double> center = [0.0, 0.0, 0.0];
  for (var p in points) {
    center[0] += p[0];
    center[1] += p[1];
    center[2] += p[2];
  }
  center = center.map((v) => v / n).toList();
  // Compute distances from center
  List<double> distances = points.map((p) =>
      sqrt(pow(p[0] - center[0], 2) + pow(p[1] - center[1], 2) + pow(p[2] - center[2], 2))
  ).toList();
  distances.sort();
  double r = quantile(distances, 0.95);
  return (4 / 3) * pi * pow(r, 3);
}

/// Computes magnitude (Euclidean norm) for each row of a list of lists.
List<double> computeMagnitude(List<List<double>> vectors) {
  return vectors.map((v) => sqrt(v.map((x) => x * x).reduce((a, b) => a + b))).toList();
}

/// Simple peak detection: returns indices of peaks where value is greater than its immediate neighbors
List<int> findPeaks(List<double> values, double threshold) {
  List<int> peaks = [];
  for (int i = 1; i < values.length - 1; i++) {
    if (values[i] > threshold &&
        values[i] > values[i - 1] &&
        values[i] > values[i + 1]) {
      peaks.add(i);
    }
  }
  return peaks;
}

/// Calculates gait features: steps, cadence, and returns detected peak indices.
Map<String, dynamic> calculateGaitFeatures(List<List<double>> accData, {int frequency = 50}) {
  List<double> mags = computeMagnitude(accData);
  double m = mean(mags);
  double s = std(mags);
  double threshold = m + s;
  List<int> peaks = findPeaks(mags, threshold);
  int steps = peaks.length;
  double durationMinutes = (accData.length / frequency) / 60.0;
  double cadence = durationMinutes > 0 ? steps / durationMinutes : 0;
  return {'steps': steps, 'cadence': cadence, 'peaks': peaks};
}

/// Computes skewness and kurtosis.
Map<String, double> calculateStatisticalFeatures(List<double> data) {
  int n = data.length;
  if (n == 0) return {'skewness': 0, 'kurtosis': 0};
  double m = mean(data);
  double s = std(data);
  double skewness = data.map((x) => pow((x - m) / s, 3)).reduce((a, b) => a + b) / n;
  double kurtosis = data.map((x) => pow((x - m) / s, 4)).reduce((a, b) => a + b) / n - 3;
  return {'skewness': skewness, 'kurtosis': kurtosis};
}

/// Stub for Total Harmonic Distortion (THD); a proper FFT-based implementation is recommended.
double calculateHarmonics(List<double> data) {
  // For now, return 0.
  return 0;
}

/// Detrends a signal by subtracting its mean.
List<double> detrend(List<double> data) {
  double m = mean(data);
  return data.map((x) => x - m).toList();
}

/// Cumulative trapezoidal integration.
List<double> cumTrapz(List<double> data, double dt) {
  List<double> integrated = List.filled(data.length, 0.0);
  for (int i = 1; i < data.length; i++) {
    integrated[i] = integrated[i - 1] + ((data[i - 1] + data[i]) / 2) * dt;
  }
  return integrated;
}

/// Calculates average gait velocity and residual step length from a 1D acceleration signal.
Map<String, double> calculateVelocityAndResidual(List<double> accSignal, {int frequency = 50}) {
  double dt = 1 / frequency;
  List<double> accDetrended = detrend(accSignal);
  List<double> velocity = cumTrapz(accDetrended, dt);

  List<int> peaks = findPeaks(accSignal, mean(accSignal) + std(accSignal));
  if (peaks.length < 2) return {'avgVelocity': 0, 'residualStepLength': 0};

  List<double> stepLengths = [];
  for (int i = 1; i < peaks.length; i++) {
    int start = peaks[i - 1];
    int end = peaks[i];
    // Compute displacement over the step interval.
    double displacement = 0;
    for (int j = start; j < end - 1; j++) {
      displacement += ((velocity[j] + velocity[j + 1]) / 2) * dt;
    }
    stepLengths.add(displacement);
  }
  double avgStepLength = mean(stepLengths);
  double residual = stepLengths.map((v) => (v - avgStepLength).abs()).reduce((a, b) => a + b) / stepLengths.length;
  double totalDisplacement = velocity.last;
  double totalTime = accSignal.length * dt;
  double avgVelocity = totalTime > 0 ? totalDisplacement / totalTime : 0;
  return {'avgVelocity': avgVelocity, 'residualStepLength': residual};
}

/// Calculates average step time and its residual from a 1D acceleration signal.
Map<String, double> calculateStepTimeFeatures(List<double> accSignal, {int frequency = 50}) {
  double dt = 1 / frequency;
  List<int> peaks = findPeaks(accSignal, mean(accSignal) + std(accSignal));
  if (peaks.length < 2) return {'avgStepTime': 0, 'residualStepTime': 0};
  List<double> stepIntervals = [];
  for (int i = 1; i < peaks.length; i++) {
    stepIntervals.add((peaks[i] - peaks[i - 1]) * dt);
  }
  double avgStepTime = mean(stepIntervals);
  double residual = stepIntervals.map((t) => (t - avgStepTime).abs()).reduce((a, b) => a + b) / stepIntervals.length;
  return {'avgStepTime': avgStepTime, 'residualStepTime': residual};
}

/// Cumulative integration to compute mean and variance of velocity signals.
Map<String, double> calculateVelocityFeatures(Map<String, List<double>> accData, {int frequency = 50}) {
  double dt = 1 / frequency;
  List<double> vx = cumTrapz(accData['acc_x']!, dt);
  List<double> vy = cumTrapz(accData['acc_y']!, dt);
  List<double> vz = cumTrapz(accData['acc_z']!, dt);
  return {
    'meanX': mean(vx),
    'varX': std(vx) * std(vx),
    'meanY': mean(vy),
    'varY': std(vy) * std(vy),
    'meanZ': mean(vz),
    'varZ': std(vz) * std(vz),
  };
}

/// Cumulative integration for angular velocity signals.
Map<String, double> calculateAngularVelocityFeatures(Map<String, List<double>> gyroData, {int frequency = 50}) {
  double dt = 1 / frequency;
  List<double> wx = cumTrapz(gyroData['gyro_x']!, dt);
  List<double> wy = cumTrapz(gyroData['gyro_y']!, dt);
  List<double> wz = cumTrapz(gyroData['gyro_z']!, dt);
  return {
    'meanX': mean(wx),
    'varX': std(wx) * std(wx),
    'meanY': mean(wy),
    'varY': std(wy) * std(wy),
    'meanZ': mean(wz),
    'varZ': std(wz) * std(wz),
  };
}

/// Calculates all combined high-level features from a segmented window of data.
/// Assumes that `df` is a list of SensorRow maps representing one window.
Map<String, double> calculateCombinedFeatures(List<SensorRow> df) {
  // Extract smartphone accelerometer data as list of [x, y, z]
  List<List<double>> smartphoneAcc = df.map((row) => [
        (row['acc_x_smartphone'] as num).toDouble(),
        (row['acc_y_smartphone'] as num).toDouble(),
        (row['acc_z_smartphone'] as num).toDouble()
      ]).toList();

  var gait = calculateGaitFeatures(smartphoneAcc, frequency: 50);
  int steps = gait['steps'];
  double cadence = gait['cadence'];

  // Use smartphone acc_z for velocity and step timing features.
  List<double> smartphoneAccZ = df.map((row) => (row['acc_z_smartphone'] as num).toDouble()).toList();
  var velRes = calculateVelocityAndResidual(smartphoneAccZ, frequency: 50);
  var stepTimes = calculateStepTimeFeatures(smartphoneAccZ, frequency: 50);

  // For velocity features over all axes (smartphone)
  List<double> accX = df.map((row) => (row['acc_x_smartphone'] as num).toDouble()).toList();
  List<double> accY = df.map((row) => (row['acc_y_smartphone'] as num).toDouble()).toList();
  List<double> accZ = smartphoneAccZ;
  Map<String, List<double>> smartphoneAccMap = {
    'acc_x': accX,
    'acc_y': accY,
    'acc_z': accZ,
  };
  Map<String, double> velocityFeats = calculateVelocityFeatures(smartphoneAccMap, frequency: 50);

  // Gyroscope-based sway features for smartphone
  List<double> gyroX = df.map((row) => (row['gyro_x_smartphone'] as num).toDouble()).toList();
  List<double> gyroY = df.map((row) => (row['gyro_y_smartphone'] as num).toDouble()).toList();
  List<double> gyroZ = df.map((row) => (row['gyro_z_smartphone'] as num).toDouble()).toList();
  double XYSway = calculateConfidenceEllipseArea(gyroX, gyroY);
  double YZSway = calculateConfidenceEllipseArea(gyroY, gyroZ);
  double XZSway = calculateConfidenceEllipseArea(gyroX, gyroZ);
  double swayVolume = calculateSwayVolume(gyroX, gyroY, gyroZ);

  // Additional statistical and frequency features on acc_z (smartphone)
  double freqRatio = 0; // You would implement a Welch method here.
  double bandPower = smartphoneAccZ.reduce((a, b) => a + b); // Stub
  double snr = smartphoneAccZ.isEmpty ? 0 : 10 * log( (pow(mean(smartphoneAccZ),2)) / (pow(std(smartphoneAccZ),2)) ) / ln10;
  var stats = calculateStatisticalFeatures(smartphoneAccZ);
  double thd = calculateHarmonics(smartphoneAccZ);

  // Smartwatch features: accelerometer velocity
  List<double> swAccX = df.map((row) => (row['acc_x_smartwatch'] as num).toDouble()).toList();
  List<double> swAccY = df.map((row) => (row['acc_y_smartwatch'] as num).toDouble()).toList();
  List<double> swAccZ = df.map((row) => (row['acc_z_smartwatch'] as num).toDouble()).toList();
  Map<String, List<double>> swAccMap = {
    'acc_x': swAccX,
    'acc_y': swAccY,
    'acc_z': swAccZ,
  };
  Map<String, double> velocityFeatsSW = calculateVelocityFeatures(swAccMap, frequency: 50);

  // Smartwatch features: gyroscope angular velocity
  List<double> swGyroX = df.map((row) => (row['gyro_x_smartwatch'] as num).toDouble()).toList();
  List<double> swGyroY = df.map((row) => (row['gyro_y_smartwatch'] as num).toDouble()).toList();
  List<double> swGyroZ = df.map((row) => (row['gyro_z_smartwatch'] as num).toDouble()).toList();
  Map<String, List<double>> swGyroMap = {
    'gyro_x': swGyroX,
    'gyro_y': swGyroY,
    'gyro_z': swGyroZ,
  };
  Map<String, double> angularVelocityFeats = calculateAngularVelocityFeatures(swGyroMap, frequency: 50);

  // Gyroscope-based sway features for smartwatch
  double XYSwaySW = calculateConfidenceEllipseArea(swGyroX, swGyroY);
  double YZSwaySW = calculateConfidenceEllipseArea(swGyroY, swGyroZ);
  double XZSwaySW = calculateConfidenceEllipseArea(swGyroX, swGyroZ);
  double swayVolumeSW = calculateSwayVolume(swGyroX, swGyroY, swGyroZ);

  return {
    // Smartphone features
    'steps_smartphone': steps.toDouble(),
    'cadence_smartphone': cadence,
    'avg_velocity_smartphone': velRes['avgVelocity']!,
    'residual_step_length_smartphone': velRes['residualStepLength']!,
    'avg_step_time_smartphone': stepTimes['avgStepTime']!,
    'residual_step_time_smartphone': stepTimes['residualStepTime']!,
    'XY_sway_area_smartphone': XYSway,
    'YZ_sway_area_smartphone': YZSway,
    'XZ_sway_area_smartphone': XZSway,
    'sway_volume_smartphone': swayVolume,
    'frequency_ratio_smartphone': freqRatio,
    'band_power_smartphone': bandPower,
    'signal_noise_ratio_smartphone': snr,
    'skewness_smartphone': stats['skewness']!,
    'kurtosis_smartphone': stats['kurtosis']!,
    'total_harmonic_distortion_smartphone': thd,
    'velocity_mean_X_smartphone': velocityFeats['meanX']!,
    'velocity_variance_X_smartphone': velocityFeats['varX']!,
    'velocity_mean_Y_smartphone': velocityFeats['meanY']!,
    'velocity_variance_Y_smartphone': velocityFeats['varY']!,
    'velocity_mean_Z_smartphone': velocityFeats['meanZ']!,
    'velocity_variance_Z_smartphone': velocityFeats['varZ']!,
    // Smartwatch features
    'XY_sway_area_smartwatch': XYSwaySW,
    'YZ_sway_area_smartwatch': YZSwaySW,
    'XZ_sway_area_smartwatch': XZSwaySW,
    'sway_volume_smartwatch': swayVolumeSW,
    'velocity_mean_X_smartwatch': velocityFeatsSW['meanX']!,
    'velocity_variance_X_smartwatch': velocityFeatsSW['varX']!,
    'velocity_mean_Y_smartwatch': velocityFeatsSW['meanY']!,
    'velocity_variance_Y_smartwatch': velocityFeatsSW['varY']!,
    'velocity_mean_Z_smartwatch': velocityFeatsSW['meanZ']!,
    'velocity_variance_Z_smartwatch': velocityFeatsSW['varZ']!,
    'angular_velocity_mean_X_smartwatch': angularVelocityFeats['meanX']!,
    'angular_velocity_variance_X_smartwatch': angularVelocityFeats['varX']!,
    'angular_velocity_mean_Y_smartwatch': angularVelocityFeats['meanY']!,
    'angular_velocity_variance_Y_smartwatch': angularVelocityFeats['varY']!,
    'angular_velocity_mean_Z_smartwatch': angularVelocityFeats['meanZ']!,
    'angular_velocity_variance_Z_smartwatch': angularVelocityFeats['varZ']!,
  };
}

/// Groups data by 'window_id' and computes features for each window.
List<Map<String, dynamic>> calculateFeatures(List<SensorRow> data) {
  // Group by window_id.
  Map<int, List<SensorRow>> groups = {};
  for (var row in data) {
    int windowId = row['window_id'] as int;
    groups.putIfAbsent(windowId, () => []).add(row);
  }
  List<Map<String, dynamic>> features = [];
  groups.forEach((windowId, rows) {
    Map<String, double> feats = calculateCombinedFeatures(rows);
    feats['window_id'] = windowId.toDouble();
    features.add(feats);
  });
  return features;
}

/// The full sensor data pipeline.
List<Map<String, dynamic>> sensorDataPipeline(List<SensorRow> rawData) {
  var preprocessed = loadAndPreprocessData(rawData);
  var cleaned = normalizeAndRemoveOutliers(preprocessed);
  var segmented = segmentData(cleaned);
  return calculateFeatures(segmented);
}
